### PR TITLE
Implement diesel compatibility

### PR DESCRIPTION
**Pull request by Kijewski**
This PR allows using CompactString as representation for [diesel](https://diesel.rs/) text columns.

### CODE REVIEW

**Comment by ParkMyCar** [comment]
Thank you for the PR! I'm going to release `v0.8.0-beta`, but we'll target to get trait impls for database crates in either a `v0.8.0-beta.1` or just `v0.8.0`, all dependent on https://github.com/ParkMyCar/compact_str/issues/304

File: tests/diesel/src/models.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+pub struct Todo {
+    pub id: i32,
+    pub title: CompactString,
+    pub done: bool,
+}
+
+#[derive(Insertable)]
+#[diesel(table_name = crate::schema::todos)]
+pub struct NewTodo<'a> {
+    pub title: &'a str,
```
Comment (NobodyXu) [inline tests/diesel/src/models.rs:16]: Why not make this

```suggestion
    pub title: CompactString,
```

to test serialization?

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: tests/diesel/src/models.rs (hunk 1 above)
Comment (Kijewski) [reply tests/diesel/src/models.rs]: Ah, yes, you're right!

File: tests/diesel/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+use tempfile::tempdir;
+
+use crate::models::{
+    NewTodo,
+    Todo,
+};
+use crate::schema::todos;
+use crate::schema::todos::dsl::*;
+
+const TITLE: &str = "Say hello!";
```
Comment (NobodyXu) [inline tests/diesel/src/lib.rs:18]: Could make it a `CompactString`:

```suggestion
const TITLE: CompactString = CompactString::from_static_str("Say hello!");
```

Kind of testing the static str impl and comparison with inline impl.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: tests/diesel/src/lib.rs (hunk 2 above)
Comment (Kijewski) [reply tests/diesel/src/lib.rs]: Good idea!

**Review by NobodyXu — APPROVED** [review]

**Comment by ParkMyCar** [comment]
Overall this looks good to me, what do you think about moving `tests/diesel` to `examples/diesel` though? Reason being, that's where all of our current examples of how to use optional features live.

Also thanks to @NobodyXu for giving this a once over!

File: tests/diesel/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+use crate::schema::todos::dsl::*;
+
+const TITLE: CompactString = CompactString::from_static_str("Say hello!");
+
+#[test]
+fn diesel_roundtrip() -> Result<(), Box<dyn std::error::Error>> {
+    // create and open a temporary database
+    let tempdir = tempdir()?;
+    let db_path = tempdir.path().join("compact_str.sqlite3");
+    let db_path = format_compact!("file://{}", db_path.to_str().unwrap());
```
Comment (ParkMyCar) [inline tests/diesel/src/lib.rs:28]: Love the use of the macro!

**Comment by Kijewski** [comment]
> what do you think about moving tests/diesel to examples/diesel though? Reason being, that's where all of our current examples of how to use optional features live.

I was not sure if how I am using diesel is best practice, and didn't want to call it an example because of that. But I guess it does not matter all that much. :) I move the code into `examples` and rebased the PR.

**Review by ParkMyCar — APPROVED** [review]

**Comment by ParkMyCar** [comment]
Fantastic, thanks @Kijewski!

**Pull request merged**